### PR TITLE
Fix CUDA code

### PR DESCRIPTION
**Pull request by gridley**
Greetings,

Here is a start on getting gloriously speedy CUDA to work once again. Under this commit, I have updated several method names referred to within the GPUSolver class, which have as of three years ago been changed in the parent class, Solver. Currently, actually running the code causes a hang once the simulation gets to the eigenvalue calculation iterations. Working on fixing that! I went ahead and made the PR in case I'm making any glaring errors.

### CODE REVIEW

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: Dont worry about chi spectrum too much, uniform starting guess is fine

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: _sin_thetas for the array name since it's global ?

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: Flipped bwd and fwd I think. 
We should rename _track_in and _track_out with _track_fwd and _track_bwd or something the other solvers use.
What do you think ?

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: see docstring in CPUSolver

File: src/accel/cuda/clone.cu
Comment (GiudGiud) [inline src/accel/cuda/clone.cu]: transfer_flux_in/out are supposed to be booleans to indicate whether the flux should be transferred to the next track

Basically they're NOT(BC==VACUUM).
I think what you wrote works (VACUUM = 0)

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: num_FSRs

File: sample-input/benchmarks/c5g7/c5g7-2d-gpu.py
Comment (GiudGiud) [inline sample-input/benchmarks/c5g7/c5g7-2d-gpu.py]: Is there routines to tally Cmfd currents in GPUSolver ?

**Review by GiudGiud — COMMENTED** [review]
Thanks for the PR! Looks like the structures have diverged quite a bit since then
Good luck with getting the rest of the solver to work

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Wait, you're preceding global variables with underscores? The other variables in constant memory also have no preceding underscore, which are at the top of this file.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Ah, thanks! Guess I should have read the documentation rather than just the code xD

Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: OK, this should do what it's supposed to now.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Are you sure here? num_FSRs is a length 1 array here so that it could be placed in GPU constant memory, which is why you access it like a pointer. Although, with this tiny example program, it seems that maybe in Will Boyd's day, __constant__ memory had to consist of arrays or pointers but now it can be values? I'm going to try changing these all to values rather than length one arrays.

```cuda
__constant__ unsigned asdf = 5;

__global__ void printSecretMessage()
{
  printf("Waddup %i\n", asdf);
}

int main()
{
  printSecretMessage<<<1,32>>>();
  cudaDeviceSynchronize();
}
```

File: sample-input/benchmarks/c5g7/c5g7-2d-gpu.py
Comment (gridley) [reply sample-input/benchmarks/c5g7/c5g7-2d-gpu.py]: Uh, definitely not. Removing that...

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: oh I meant "global" as in a class attribute as opposed to a local function-scope variable

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: Well if it's an array you're right!
But let's make it a value.

Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Your wish is my command! (done) I imagine this has to do with some old version of CUDA. I checked cudaMemcpyToSymbol with plain old values and it seems to work fine.

**Comment by gridley** [comment]
So, it seems that the main thing holding this back from working right now is Thrust not doing its job, see [this](https://groups.google.com/forum/#!topic/thrust-users/LneZ66OZ_yg) post I made to their mailing list for an explanation. I'm thinking about just implementing our own device_vector data structure since it's not crazy hard...

File: config.py
Comment (GiudGiud) [inline config.py:384]: It's not a nvcc flag, it's a compiler flag to help with debugging large functions. I read about it there : [link](https://stackoverflow.com/questions/14666665/trying-to-understand-gcc-option-fomit-frame-pointer
) and decided to add it.

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: add nu to the docstring too

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: remove coments

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: use log_printf

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: log_printf

File: src/accel/cuda/GPUSolver.h
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.h:128]: any technical reason you're switching to double precision ?
It doesnt make much of a difference here

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:1238]: great idea !

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:1100]: put an error log for now

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: ok done in a later commit

**Comment by GiudGiud** [comment]
Looks good!
Let me know when it's ready for merging.
I ll test it for accuracy/correctness then.

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: Ok looking at the VectorizedSolver which also uses the previous format, getTrackIn is for the Bwd direction

Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: And looking at DeviceTrack.h, which disagrees,
** Index of the next Track when traveling along this Track in the "forward" direction. */
  int _track_in;

Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: looking at GPUSolver.cu line 506, you can see that _track_in is used in the backward direction.

SO
- I was wrong to ask for a change here
- deviceTrack docstrings are wrong and need to be modified

File: src/accel/cuda/clone.cu
Comment (GiudGiud) [reply src/accel/cuda/clone.cu]: In and Out are flipped. In is backward. 

This whole mess shows we should change notation to something more explicit

File: config.py
Comment (gridley) [reply config.py:384]: Ah, OK, thanks for catching that. It seems that nvcc doesn't allow this option, so, I've changed the script to now only apply it if nvcc isn't used.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: aight

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: done

File: src/accel/cuda/GPUSolver.h
Comment (gridley) [reply src/accel/cuda/GPUSolver.h:128]: Yes, actually. double now has to be used here. Because GPUSolver inherits from Solver, these have to be returning doubles, since that what Solver.h has specified. Otherwise, they won't override their parent class's methods. I had an issue with these not overriding at first, which is why I added the override keywords to give errors in case anyone changes the return value and unknowingly starts making GPUSolver call methods from Solver, which would make it not work.

**Comment by gridley** [comment]
OK, cool, so, the last commit addresses all those comments. It will probably be a few more days (hopefully, do time estimates on coding ever work out?) before this works.

File: src/accel/cuda/clone.cu
Comment (gridley) [reply src/accel/cuda/clone.cu]: Hm, that is some confusing notation. So, I just changed _next_in to _next_bwd and vice versa for the other. Looks far easier to understand now.

File: src/accel/cuda/GPUSolver.h
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.h:128]: Ok works for me! We may want to change all the solvers to return FP_PRECISION since OpenMOC should be run in single precision most of the time.

**Comment by gridley** [comment]
Hey, so, does initializeSolver not get called now for some reason @GiudGiud? This is creating some issues with GPUSolver, which assumes the initializeSolver method will get called. It seems that you can attempt to run GPUSolver's computeEigenvalue without initializing it, which has confused me very much because I for some reason assumed initializeSolver would be called in the constructor.

**Comment by gridley** [comment]
OK, so, anyways, now running the computeEigenvalue method in GPUSolver now actually runs and does transport sweeps, but seems to numerically blow up in a few iterations. This will take some effort to fix.

**Comment by GiudGiud** [comment]
InitializeSolver is not run on the C++ side by the computeFlux/Source/Eigenvalue routines or the constructor. It was added on the Python side, it was added when Colin Josey created the Krylov solver I think, and he needed a bit more control on the solver from Python

In initializeSolver you essentially have a couple routines for each part of the solver. These are called separately in the computeFlux/Source/Eig routines.

Good luck with the bug search. I'd set track angular fluxes to 0 / do a vacuum boundary case and see if it works fine there first.

**Comment by gridley** [comment]
OK, cool. So, I think the issue is in GPUExpEvaluator. In particular, it seems that in the past, there was not a matrix of ExpEvaluators in solver for each of the polar and azimuthal angles. What is the purpose of this?

**Comment by GiudGiud** [comment]
A sin(theta) term comes up in the exponential terms (tau = XSt * length_2D / sin(theta)). 
In 3D MOC the polar angles are slightly adjusted for each azimuthal angle to make the tracks cyclic. 
So it's better to use a different ExpEval per azimuthal angle. 

Note that we use rational fractions not expEvaluators with interpolation tables now. Look at tallyScalarFlux in CPUSolver.cpp and the call to expF1Fractional(tau, &exp). Making that change for GPU would also work and may be easier than fixing interpolation tables.

**Comment by gridley** [comment]
It is with great pleasure I present the following table for results on C5G7 with 32 azimuthal angles, and everything else set to default values from the c5g7-2d.py input. This is on my laptop with an  Intel Core i7-8750H and Nvidia Quadro P1000.

```
      Time      Keff
     -------------------
CPU  | 648 s  | 1.186703
GPU  | 35  s  | 1.186700
```

The PR is ready after review, aside from the fact that some regression test is failing about IHM multisim, and I have absolutely no clue why. By the way, turning on Cmfd seems to give my CPU solver a different result, about 100 pcm different from the case here. That's why I was unable to get GPU to match CPU.

File: sample-input/benchmarks/c5g7/c5g7-2d.py
Comment (GiudGiud) [inline sample-input/benchmarks/c5g7/c5g7-2d.py]: default to False

File: src/Solver.cpp
Comment (GiudGiud) [inline src/Solver.cpp]: no longer needed with proposed change, see c5g7-2d.py comments

File: sample-input/cosine-source/cosine-source.py
Comment (GiudGiud) [inline sample-input/cosine-source/cosine-source.py:117]: same

File: sample-input/benchmarks/c5g7/c5g7-2d.py
Comment (GiudGiud) [inline sample-input/benchmarks/c5g7/c5g7-2d.py]: Get rid of this line and swap line 1370 and 1371 in Solver.cpp so that FSRs are initialized before materials.

File: src/Solver.h
Comment (GiudGiud) [inline src/Solver.h]: remove

File: src/Solver.cpp
Comment (GiudGiud) [reply src/Solver.cpp]: also it may be what s causing the test failure for multi-simulations

File: src/Solver.cpp
Comment (GiudGiud) [inline src/Solver.cpp]: Remove those, this is already checked in the call to geometry.getNumEnergyGroups()

File: src/accel/cuda/GPUSolver.h
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.h]: , on device! -> on the device

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:4]: any reason all of these arent in the header ?

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:32]: use log_printf
Also this should be a class function of dev_material

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: most probably negligible effect to use the value rather than the pointer.

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: delete comment, this an initialization routine, that we dont use because taking the Chi spectrum as a first guess isnt actually a good idea (will crash your CMFD), so it wont need to go faster

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: A 0.5% speedup is going to be very very hard to measure
If you want to keep the comment, fix next_fwd_is_bwd that appears twice

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: Here ? Not sure what you mean

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:659]: add computeKeff argument in param

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: remove

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: computekeff -> computing_fission_norm is a more explicit name

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:1144]: do we want to keep these getLastCudaError calls everywhere ?

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: remove

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: remove

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: remove

File: src/accel/cuda/dev_exponential.cu
Comment (GiudGiud) [inline src/accel/cuda/dev_exponential.cu]: this function call should probably be inlined. Could you please look if that gets us anything ?

File: src/accel/cuda/dev_exponential.cu
Comment (GiudGiud) [inline src/accel/cuda/dev_exponential.cu]: Add comment
//TODO Redo fractional approximation for 1-exp(-x) rather than (1-exp(-x)) / x

Comment (GiudGiud) [reply src/accel/cuda/dev_exponential.cu]: Right now, the appxomiation error is non-uniform because it is multiplied by x, so smaller error below 1 and bigger above

**Review by GiudGiud — CHANGES_REQUESTED** [review]
Great PR!
It runs well and pretty fast. Thanks for taking care of this

I'll rework the timer to show stats for the GPU once this is merged.

**Comment by GiudGiud** [comment]
For the disagreement (100pcm) with the solution using CMFD, that's because that one is typically much more converged. You want to run a very tight convergence criterion to get that level of convergence without CMFD. For this case, 1e-6 gets you 1 pcm away

File: src/Solver.cpp
Comment (gridley) [reply src/Solver.cpp]: Gee, I hope! Let's see...

File: sample-input/benchmarks/c5g7/c5g7-2d.py
Comment (gridley) [reply sample-input/benchmarks/c5g7/c5g7-2d.py]: Hm. It seems that if I remove initializeSolver when the GPUSolver is being used, the number of groups used in the calculation doesn't get communicated to the GPU. Luckily that error now gets caught in GPUSolver before havoc ensues.

File: src/accel/cuda/dev_exponential.cu
Comment (gridley) [reply src/accel/cuda/dev_exponential.cu]: Ah dang, silly me for not doing that. So, slapping on a __forceinline__ takes C5G7 with 64 azimuthals from 50s to 34s total run time without plotting. Shoo wee, that is fast. I've read the nvcc compiler automatically inlines stuff, so I'm amazed it wasn't doing that before.

File: sample-input/cosine-source/cosine-source.py
Comment (gridley) [reply sample-input/cosine-source/cosine-source.py:117]: see next comment

File: src/accel/cuda/dev_exponential.cu
Comment (GiudGiud) [reply src/accel/cuda/dev_exponential.cu]: nice, 30% performance increase in 5 minutes of work, the way to go
When it's in different source files you need some link time optimization to get inlining. I dont know much about nvcc but it s probably not done by default

**Review by GiudGiud — COMMENTED** [review]
Couple comments were folded by github in last review. Should show up now that the others were resolved.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu:1144]: Hm, I think it would be good to have them. Does any compiler macro variable get defined when --debug is used? Then I could only enable them with ifdef DEBUG or the like.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu:4]: nope, that's just the way I found it. I agree they should be in the header.

Comment (gridley) [reply src/accel/cuda/GPUSolver.cu:4]: Actually, it seems that for CUDA, you have to put __constant__ variables in the implementation file and not the header, otherwise you get multiple definition errors.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: removed!

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: removed

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: So, it's a question of code expressiveness and not really a big deal, so I removed the comment. It's there because you end up multiplying something by a boolean rather than just using an if statement. I think just using the boolean as is is fine, so the comment is removed.

File: src/accel/cuda/GPUSolver.cu
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: done

**Comment by gridley** [comment]
Ah, OK, stuff should be fixed now.

**Review by GiudGiud — APPROVED** [review]

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: Ok. I agree it should work either way. Thanks for the insight though

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu:4]: Ok good to know. Thanks for checking!

File: src/accel/cuda/GPUSolver.cu
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu:1144]: It doesnt. We'll have to think about this. Current abstraction is using those log_printf(DEBUG)

**Review by GiudGiud — APPROVED** [review]
Merging! Works great, is really fast, thanks for the PR!

**Pull request merged**
